File: TFTP/Server/src/main/java/impl/tftp/PacketEncoderDecoder.java
package bgu.spl171.net.impl.tftp;


import java.util.Arrays;

import bgu.spl171.net.api.MessageEncoderDecoder;

public class PacketEncoderDecoder implements MessageEncoderDecoder<Command>{

	private byte[] opCodeArr = new byte[2];
	private byte[] content = new byte[1 << 10];
	private int conIndex = 0;
    private int opCodeIndex = 0;
    private short opCode;
    private short dataSize;
    
    
	@Override
	public Command decodeNextByte(byte nextByte) {
		if (opCodeIndex<2){
			opCodeArr[opCodeIndex++] = nextByte;
			if (opCodeIndex == 2){
				opCode = bytesToShort(opCodeArr);
				if (opCode==6){ // Dir
					clear();
					return new CommandDir();
				}
				if (opCode==10){ // Disc
					clear();
					return new CommandDisc();
				}
			}
		}
		else{
			Command ans;
			switch (opCode){
				case(1): // Read
					if (nextByte == '\0'){
						ans = new CommandRead(new String(content,0,conIndex));
						clear();
						return ans;
					}
					pushByte(nextByte);
					break;
				case(2): // Write
					if (nextByte == '\0'){
						ans = new CommandWrite(new String(content,0,conIndex));
						clear();
						return ans;
					}
					pushByte(nextByte);
					break;
				case(3): // Data
					if (conIndex<2){
						pushByte(nextByte);
						if (conIndex == 2)
							dataSize = bytesToShort(Arrays.copyOfRange(content, 0, 2));
					}
					else{
						pushByte(nextByte);
						if (conIndex == dataSize + 4){
							ans = new CommandData(dataSize,bytesToShort(Arrays.copyOfRange(content, 2, 4)),Arrays.copyOfRange(content, 4, conIndex));
							clear();
							return ans;
						}
					}
					break;
				case(4): // Ack
					pushByte(nextByte);
					if (conIndex == 2){
						ans = new CommandAck(bytesToShort(Arrays.copyOfRange(content, 0, 2)));
						clear();
						return ans;
					}
					break;
				case(5): // Error
					if (nextByte == '\0'){
						ans = new CommandError(bytesToShort(Arrays.copyOfRange(content, 0, 2)),new String(Arrays.copyOfRange(content, 2, conIndex)));
						clear();
						return ans;
					}
					pushByte(nextByte);
					break;
				case(7): // Log
					if (nextByte == '\0'){
						ans = new CommandLog(new String(content,0,conIndex));
						clear();
						return ans;
					}
					pushByte(nextByte);
					break;
				case(8): // Del
					if (nextByte == '\0'){
						ans = new CommandDel(new String(content,0,conIndex));
						clear();
						return ans;
					}
					pushByte(nextByte);
					break;
				case(9): // Bcast
					if (nextByte == '\0'){
						ans = new CommandBcast(content[0],new String(content,1,conIndex-1));
						clear();
						return ans;
					}
					pushByte(nextByte);
					break;
				default:
					break;
					
			}
		}
		return null;
	}
	
	private void pushByte(byte nextByte) {
        if (conIndex >= content.length)
        	content = Arrays.copyOf(content, conIndex * 2);
        content[conIndex++] = nextByte;
    }
	
	private void clear(){
		conIndex = 0;
		opCodeIndex = 0;
	}

	@Override
	public byte[] encode(Command message) {
		byte[] temp;
		byte[] opArr = shortToBytes(message.opCode);
		switch(message.opCode){
			case(1): // Read
				temp = (((CommandRead) message).getStr() + '\0').getBytes();
				break;
			case(2): // Write
				temp = (((CommandWrite) message).getStr() + '\0').getBytes();
				break;
			case(3): // Data
				byte[] size = shortToBytes(((CommandData) message).getSize());
				byte[] blockNum = shortToBytes(((CommandData) message).getBlockNumber());
				byte[] data = ((CommandData) message).getData();
				temp = new byte[data.length + size.length + blockNum.length];
				System.arraycopy(size, 0, temp, 0, 2);
				System.arraycopy(blockNum, 0, temp, 2, 2);
				System.arraycopy(data, 0, temp, 4, data.length);
				break;
			case(4): // Ack
				temp = shortToBytes(((CommandAck) message).getBlock());
				break;
			case(5): // Error
				byte[] msgStr = ((CommandError) message).getErrorMsg().getBytes();
				byte[] errCode = shortToBytes(((CommandError) message).getErrorCode());
				temp = new byte[msgStr.length + errCode.length];
				System.arraycopy(errCode, 0, temp, 0, errCode.length);
				System.arraycopy(msgStr, 0, temp, 2, msgStr.length);
				break;
			case(6): // Dir
				temp = new byte[0];
				break;
			case(7): // Log
				temp = (((CommandLog) message).getStr() + '\0').getBytes();		
				break;
			case(8): // Del
				temp = (((CommandDel) message).getStr() + '\0').getBytes();
				break;
			case(9): // Bcast
				byte[] name = (((CommandBcast) message).getName() + '\0').getBytes();
				temp = new byte[name.length + 1];
				temp[0] = ((CommandBcast) message).getType();
				System.arraycopy(name, 0, temp, 1, name.length);
				break;
			default: // Disc
				temp = new byte[0];
				break;
		}
		byte[] ans = new byte[opArr.length + temp.length];
		System.arraycopy(opArr, 0, ans, 0, opArr.length);
		System.arraycopy(temp, 0, ans, 2, temp.length);
		return ans;
	}
	
	private short bytesToShort(byte[] byteArr){
	    short result = (short)((byteArr[0] & 0xff) << 8);
	    result += (short)(byteArr[1] & 0xff);
	    return result;
	}
	
	private byte[] shortToBytes(short num)
	{
	    byte[] bytesArr = new byte[2];
	    bytesArr[0] = (byte)((num >> 8) & 0xFF);
	    bytesArr[1] = (byte)(num & 0xFF);
	    return bytesArr;
	}

}

File: TFTP/Client/src/PacketEncoderDecoder.cpp
#include "../include/PacketEncoderDecoder.h"
#include <string.h>
#include <stdio.h>

PacketEncoderDecoder::PacketEncoderDecoder():opcodeArr(), opcode(0), opcodeIndex(0), content(), conIndex(0), dataSize(0){
	opcodeArr = new char[2];
	content = new std::vector<char>();
}

PacketEncoderDecoder::PacketEncoderDecoder(const PacketEncoderDecoder& other):opcodeArr(), opcode(other.opcode), opcodeIndex(other.opcodeIndex), content(), conIndex(other.conIndex), dataSize(other.dataSize){
	opcodeArr = new char[2];
	opcodeArr[0] = other.opcodeArr[0];
	opcodeArr[1] = other.opcodeArr[1];
	content = new std::vector<char>(*other.content);
}

PacketEncoderDecoder& PacketEncoderDecoder::operator=(const PacketEncoderDecoder& other){
	if (this!=&other){
		opcode = other.opcode;
		opcodeIndex = other.opcodeIndex;
		conIndex = other.conIndex;
		dataSize = other.dataSize;
		char* temp = new char[2];
		std::vector<char>* tempVec = new std::vector<char>(*other.content);
		temp[0] = other.opcodeArr[0];
		temp[1] = other.opcodeArr[1];
		delete[] opcodeArr;
		delete content;
		opcodeArr = temp;
		content = tempVec;
	}
	return *this;
}

Command* PacketEncoderDecoder::decodeNextByte(char nextByte){
	if (opcodeIndex < 2){
		opcodeArr[opcodeIndex++] = nextByte;
		if (opcodeIndex == 2)
			opcode = bytesToShort(opcodeArr);
	}
	else{
		Command *ans;
		switch (opcode){
			case 3: //data
				if (conIndex < 2){
					pushByte(nextByte);
					if (conIndex == 2){
						char* dataSizeArr = new char[2];
						dataSizeArr[0] = content->at(0);
						dataSizeArr[1] = content->at(1);
						dataSize = bytesToShort(dataSizeArr);
						delete[] dataSizeArr;
					}
				}
				else{
					pushByte(nextByte);
					if (conIndex == dataSize + 4){
						char* dataBlockNumber = new char[2];
						dataBlockNumber[0] = content->at(2);
						dataBlockNumber[1] = content->at(3);
						char* data = new char[dataSize];
						std::copy(content->begin()+4,content->begin()+conIndex,data);
						ans = new CommandData(dataSize,bytesToShort(dataBlockNumber), data);
						delete[] dataBlockNumber;
						clear();
						return ans;
					}
				}
				break;
			case 4: //ack
				pushByte(nextByte);
				if (conIndex == 2){
					char* ackBlockNumber = new char[2];
					ackBlockNumber[0] = content->at(0);
					ackBlockNumber[1] = content->at(1);
					ans = new CommandAck(bytesToShort(ackBlockNumber));
					delete[] ackBlockNumber;
					clear();
					return ans;
				}
				break;
			case 5: //error
				if (nextByte == '\0' && conIndex>1){
					char* errorCode = new char[2];
					errorCode[0] = content->at(0);
					errorCode[1] = content->at(1);
					ans = new CommandError(bytesToShort(errorCode),std::string(content->begin()+2,content->begin()+conIndex));
					delete[] errorCode;
					clear();
					return ans;
				}
				pushByte(nextByte);
				break;
			case 9: //bcast
				if (nextByte == '\0' && conIndex>0){
					ans = new CommandBcast(std::string(content->begin()+1,content->begin()+conIndex),content->at(0));
					clear();
					return ans;
				}
				pushByte(nextByte);
				break;
		}
	}
	return nullptr;
}


std::string PacketEncoderDecoder::encode(Command *message){
	std::string temp;
	char* op = new char[2];
	shortToBytes(message->getOpcode(),op);
	switch (message->getOpcode()){
		case 1:{ //read
			temp = (dynamic_cast<CommandRead*>(message))->getStr();
			break;
		}
		case 2:{ //write
			temp = (dynamic_cast<CommandWrite*>(message))->getStr();
			break;
		}
		case 3:{ //data
			char* dataSize = new char[2];
			int size = (dynamic_cast<CommandData*>(message))->getPacketSize();
			shortToBytes(size,dataSize);
			char* dataBlockNum = new char[2];
			shortToBytes((dynamic_cast<CommandData*>(message))->getBlockNumber(),dataBlockNum);
			char* data = (dynamic_cast<CommandData*>(message))->getData();
			temp.insert(0,1,dataBlockNum[1]);
			temp.insert(0,1,dataBlockNum[0]);
			temp.insert(0,1,dataSize[1]);
			temp.insert(0,1,dataSize[0]);
			for (int i=0; i<size; i++){
				temp = temp + data[i];
			}
			delete[] dataSize;
			delete[] dataBlockNum;
			break;
		}
		case 4:{ //ack
			char* ackBlockNum = new char[2];
			shortToBytes((dynamic_cast<CommandAck*>(message))->getBlockNumber(),ackBlockNum);
			temp.insert(0,1,ackBlockNum[1]);
			temp.insert(0,1,ackBlockNum[0]);
			delete[] ackBlockNum;
			break;
		}
		case 5:{ //error
			char* errorCode = new char[2];
			shortToBytes((dynamic_cast<CommandError*>(message))->getErrorCode(),errorCode);
			temp = (dynamic_cast<CommandError*>(message))->getErrorMsg();
			temp.insert(0,1,errorCode[1]);
			temp.insert(0,1,errorCode[0]);
			delete[] errorCode;
			break;
		}
		case 7:{ //login
			temp = (dynamic_cast<CommandLog*>(message))->getStr();
			break;
		}
		case 8:{ //delete
			temp = (dynamic_cast<CommandDel*>(message))->getStr();
			break;
		}
	}
	temp.insert(0,1,op[1]);
	temp.insert(0,1,op[0]);
	delete[] op;
	return temp;
}

void PacketEncoderDecoder::pushByte(char nextByte){
	content->push_back(nextByte);
	conIndex++;
}


void PacketEncoderDecoder::clear(){
	content->clear();
	conIndex = 0;
	opcodeIndex = 0;
	opcode = 0;
	dataSize = 0;

}


short PacketEncoderDecoder::bytesToShort(char* bytesArr){
    short result = (short)((bytesArr[0] & 0xff) << 8);
    result += (short)(bytesArr[1] & 0xff);
    return result;
}

void PacketEncoderDecoder::shortToBytes(short num, char* bytesArr){
    bytesArr[0] = ((num >> 8) & 0xFF);
    bytesArr[1] = (num & 0xFF);
}

PacketEncoderDecoder::~PacketEncoderDecoder(){
	delete[] opcodeArr;
	delete content;
}
